### PR TITLE
REST API auth

### LINKED ISSUE #721: Get tribe/agreements is called when just issue snapctl
It is more of a question. 
When running snapctl without any command and noticed that "get tribe/agreements" is issued from snapctl.

Is there a particular use-case for that?

I'm using " v0.10.0-beta-2-g5da30c0"

On the snapd side, the follow message is shown.

INFO[0816] API response                                  _module=_mgmt-rest index=49 method=GET status=Not Found status-code=404 url=/v1/tribe/agreements

### PR DESCRIPTION
**Pull request by tiffanyfj**
-Added password authentication for the REST API in server and client
-Updated docs
-Fixes #721

### CODE REVIEW

File: cmd/snapctl/flags.go [hunk 1]
```diff
@@ -43,6 +43,16 @@ var (
 		Name:  "running",
 		Usage: "Shows running plugins",
 	}
+	flPassword = cli.BoolFlag{
+		Name:  "password, p",
+		Usage: "Password for REST API authentication",
+	}
+	flPasswordPath = cli.StringFlag{
+		Name:   "Password, P",
```
Comment (danielscottt) [inline cmd/snapctl/flags.go]:
```go
Name: "password file, P",
```
Maybe?

Comment (danielscottt) [inline cmd/snapctl/flags.go]: Better yet, I see below that the env is called `SNAPCTL_PASS_PATH`, let's call this `"password file path, P"`.

File: cmd/snapctl/main.go [hunk 2]
```diff
@@ -88,12 +110,59 @@ func init() {
...
+		} else {
+			password = string(pass)
+		}
+		// Go to next line after password prompt
+		fmt.Println()
+	}
+	pClient.Username = username
+	pClient.Password = strings.TrimSpace(password)
+	// Send tribe agreement requests only if needed by help
+	if help {
```
Comment (danielscottt) [inline cmd/snapctl/main.go]: :+1:

Going to tag: #721 here.

File: docs/REST_API.md [hunk 3]
```diff
@@ -11,19 +11,44 @@ snap exposes a list of RESTful APIs to perform various actions. All of snap's AP
...
+```json
+{
+  "meta": {
+    "code": 200,
+    "message": "Plugin list returned",
+    "type": "plugin_list_returned",
+    "version": 1
+  },
+  "body": {}
+}%
```
Comment (danielscottt) [inline docs/REST_API.md]: nitpick: Can you remove the `%` here?

File: mgmt/rest/server.go [hunk 4]
```diff
@@ -135,6 +139,35 @@ func New(https bool, cpath, kpath string) (*Server, error) {
...
+// Auth Middleware for REST API
+func (s *Server) authMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	if s.auth {
+		_, password, ok := r.BasicAuth()
+		defer r.Body.Close()
+		err := bcrypt.CompareHashAndPassword(s.authpwd, []byte(password))
+		// If we have valid password or going to tribe/agreements endpoint
+		// go to next. tribe/agreements endpoint used for populating
+		// snapctl help page when tribe mode is turned on.
+		if (ok && err == nil) || r.RequestURI == "/v1/tribe/agreements" {
```
Comment (danielscottt) [inline mgmt/rest/server.go]: I don't think we want to give access to the `/v1/tribe/agreements` endpoint, just so `snapctl` can check on Tribe availability.  I would rather handle this on the `snapctl` side, either by removing the feature toggling all together, or printing something to the screen about being unable to authenticate to the daemon and then printing the normal help output.

File: cmd/snapctl/flags.go (hunk 1 above)
Comment (IRCody) [inline cmd/snapctl/flags.go]: Would you pass that in as --password_file_path?

Comment (geauxvirtual) [inline cmd/snapctl/flags.go]: We should come up with a better ENV name and command line flag that doesn't include PASSWORD or PASS in the name. A username and password flag on the command line should be different from an auth or credential file being passed in. Ideally, I would say snapctl should have the option of a config file that could hold this information and other config related options.

File: mgmt/rest/server.go [hunk 5]
```diff
@@ -135,6 +139,35 @@ func New(https bool, cpath, kpath string) (*Server, error) {
...
+// SetAPIAuthHashedPwd sets the API authentication password from snapd
+func (s *Server) SetAPIAuthHashedPwd(pwd []byte) {
+	s.authpwd = pwd
+}
+
+// Auth Middleware for REST API
+func (s *Server) authMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	if s.auth {
+		_, password, ok := r.BasicAuth()
+		defer r.Body.Close()
```
Comment (danielscottt) [inline mgmt/rest/server.go]: Do we only want to close Body `if s.auth`?

Comment (tiffanyfj) [inline mgmt/rest/server.go]: Nope, I forgot to move it.

**Comment by danielscottt** [comment]
LGTM - Let's get this in.

File: cmd/snapctl/config.go [hunk 6]
```diff
@@ -69,12 +91,16 @@ func getConfig(ctx *cli.Context) {
 	}
 	w := tabwriter.NewWriter(os.Stdout, 0, 8, 1, '\t', 0)
 	defer w.Flush()
+	r := pClient.GetPluginConfig(ptyp, pname, strconv.Itoa(pver))
+	if r.Err != nil {
+		fmt.Println("Error requesting info: ", r.Err)
```
Comment (candysmurf) [inline cmd/snapctl/config.go]: change fmt to log

**Comment by candysmurf** [comment]
Good work!

**Comment by danielscottt** [comment]
Ping @tiffanyfj @IRCody - Y'all ready to merge this?

**Comment by IRCody** [comment]
We were hoping to get @geauxvirtual to take a look at it also.

File: snapd.go [hunk 7]
```diff
@@ -479,14 +505,27 @@ func action(ctx *cli.Context) {
...
+		//Rest Authentication
+		if restAuth {
+			log.Info("REST API authentication is enabled")
+			r.SetAPIAuth(restAuth)
+			// restAuthPwd provided in config file
+			password := []byte(restAuthPwd)
+			log.Info("REST API authentication password is set")
+			hashedPwd, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
+			if err != nil {
+				log.Fatal("Unable to hash password from snapctl: ", err.Error)
```
Comment (geauxvirtual) [inline snapd.go]: This error message needs to be changed. snapctl plays no part in snapd starting the different modules of the system.

File: snapd.go [hunk 8]
```diff
@@ -479,14 +505,27 @@ func action(ctx *cli.Context) {
 		r.BindMetricManager(c)
 		r.BindConfigManager(c.Config)
 		r.BindTaskManager(s)
+		//Rest Authentication
+		if restAuth {
+			log.Info("REST API authentication is enabled")
+			r.SetAPIAuth(restAuth)
+			// restAuthPwd provided in config file
+			password := []byte(restAuthPwd)
+			log.Info("REST API authentication password is set")
```
Comment (geauxvirtual) [inline snapd.go]: This log line should be after the error check.

**Comment by danielscottt** [comment]
:+1:

**Comment by danielscottt** [comment]
Sorry to say it guys, but you have conflicts now.  Need to resolve and repush.  Then we can merge.

**Comment by danielscottt** [comment]
shipping.  good job guys.

**Pull request merged**